### PR TITLE
[JENKINS-31484] Translate field accesses to AttributeAccessBlock rather than PropertyAccessBlock

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-31484](https://issues.jenkins-ci.org/browse/JENKINS-31484)

@reviewbybees esp. @kohsuke

### CODE REVIEW

File: src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy [hunk 1]
```diff
@@ -785,14 +785,21 @@ class CpsTransformer extends CompilationCustomizer implements GroovyCodeVisitor
...
                 makeNode("javaThis_")
                 literal(exp.name)
             }
         } else
+        if (ref instanceof FieldNode) {
+            makeNode("attribute") {
+                loc(exp)
+                makeNode("javaThis_")
+                visit(new ConstantExpression(exp.name))
+                literal(false)
```
Comment (jglick) [inline src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy]: Really just guessing what to do here…

File: src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy [hunk 2]
```diff
@@ -612,6 +612,13 @@ class CpsTransformerTest extends AbstractGroovyCpsTest {
         ''') == [-3,2];
     }
 
+    @Issue("JENKINS-31484")
+    @Test
+    void fieldViaGetter() {
+        assert evalCPS('class C {private int x = 33}; new C().x') == 33
+        assert evalCPS('class C {private int x = 33; int getX() {2 * this.@x}}; new C().x') == 66
+        assert evalCPS('class C {private int x = 33; int getX() {2 * x}}; new C().x') == 66
```
Comment (jglick) [inline src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy]: Originally the first two were fine, but the third would go into an endless loop. Instrumenting `Next.step` indicated that accessing `x` was calling the property, which then called `getX()`  again.

**Comment by jglick** [comment]
Hmm, found another anomalous hang involving setters…

**Comment by jglick** [comment]
```diff
diff --git a/src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy b/src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy
index c0ba149..168f585 100644
--- a/src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy
+++ b/src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy
@@ -183,7 +183,7 @@ class CpsTransformer extends CompilationCustomizer implements GroovyCodeVisitor
         Expression body;
 
         // transform the body
-        parent = { e -> body=e }
+        parent = { e -> body=e; println "transformed $m.typeDescriptor to $e.text" }
         visitWithSafepoint(m.code)
 
         def params = new ListExpression();
```

shows:

```
transformed void main([Ljava.lang.String;) to b.functionCall(-1, b.constant(org.codehaus.groovy.runtime.InvokerHelper), b.constant(runScript), false, b.constant(Script1), b.localVariable(-1, args))
transformed java.lang.Object run() to b.block(b.declareVariable(1, C, c, b.new_(1, C)), b.assign(1, b.property(1, b.localVariable(1, c), b.constant(x), false), b.constant(33)), b.property(1, b.localVariable(1, c), b.constant(x), false))
transformed int getX() to b.block(b.multiply(1, b.constant(2), b.attribute(1, b.javaThis_(), b.constant(x), false)))
transformed void setX(int) to b.block(b.assign(1, b.property(1, b.this_(), b.constant(x), false), b.div(1, b.localVariable(1, x), b.constant(3))))
```

File: src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy [hunk 3]
```diff
@@ -612,6 +612,24 @@ class CpsTransformerTest extends AbstractGroovyCpsTest {
         ''') == [-3,2];
     }
 
+    @Test
+    void fieldDirect() {
+        assert evalCPS('class C {private int x = 33}; new C().x') == 33
```
Comment (jglick) [inline src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy]: always worked

File: src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy [hunk 4]
```diff
@@ -612,6 +612,24 @@ class CpsTransformerTest extends AbstractGroovyCpsTest {
...
+    }
+
+    @Issue("JENKINS-31484")
+    @Test
+    void fieldViaGetter() {
+        assert evalCPS('class C {private int x = 33; int getX() {2 * this.@x}}; new C().x') == 66
```
Comment (jglick) [inline src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy]: always worked

File: src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy [hunk 5]
```diff
@@ -612,6 +612,24 @@ class CpsTransformerTest extends AbstractGroovyCpsTest {
...
+        assert evalCPS('class C {private int x = 33; int getX() {2 * x}}; new C().x') == 66
```
Comment (jglick) [inline src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy]: now fixed

File: src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy [hunk 6]
```diff
@@ -612,6 +612,24 @@ class CpsTransformerTest extends AbstractGroovyCpsTest {
...
+    }
+
+    @Issue("JENKINS-31484")
+    @Test
+    void fieldViaSetter() {
+        assert evalCPS('class C {private int x = 0; int getX() {2 * x}; void setX(int x) {this.@x = x / 3}}; C c = new C(); c.x = 33; c.x') == 22
```
Comment (jglick) [inline src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy]: always worked

File: src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy [hunk 7]
```diff
@@ -612,6 +612,24 @@ class CpsTransformerTest extends AbstractGroovyCpsTest {
...
+        assert evalCPS('class C {private int x = 0; int getX() {2 * x}; void setX(int x) {this.x = x / 3}}; C c = new C(); c.x = 33; c.x') == 22
```
Comment (jglick) [inline src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy]: now fixed

File: src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy [hunk 8]
```diff
@@ -183,7 +183,10 @@ class CpsTransformer extends CompilationCustomizer implements GroovyCodeVisitor
         Expression body;
 
         // transform the body
-        parent = { e -> body=e }
+        parent = {
+            e -> body = e
+            // println "transformed $m.typeDescriptor to $e.text"
```
Comment (abayer) [inline src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy]: That println meant to be there?

Comment (jglick) [inline src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy]: Commented out but was invaluable during debugging so I do not want to lose it.

Comment (abayer) [inline src/main/groovy/com/cloudbees/groovy/cps/CpsTransformer.groovy]: Fair enough!

**Comment by stephenc** [comment]
:bee: AIUI

**Comment by jglick** [comment]
Found that this causes a regression in `WorkflowLibRepositoryTest.userDefinedGlobalVariable`. Back to the drawing board.

File: src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy [hunk 9]
```diff
@@ -612,6 +612,29 @@ class CpsTransformerTest extends AbstractGroovyCpsTest {
...
+    }
+
+    @Test
+    void nonField() {
+        assert evalCPS('class C extends HashMap {def read() {x * 2}; def write(x) {this.x = x / 3}}; C c = new C(); c.write(33); c.read()') == 22
```
Comment (jglick) [inline src/test/groovy/com/cloudbees/groovy/cps/CpsTransformerTest.groovy]: always worked

**Pull request merged**